### PR TITLE
oaknut: Add basic CPU feature detection

### PR DESCRIPTION
**Pull request by merryhime**
We shove all our feature detection code into a [`feature_detection` folder](https://github.com/merryhime/oaknut/tree/4b79979fcc54ca5ba6b05efc8f415f39d3d12e87/include/oaknut/feature_detection).

We want a simple interface: A user should be able to include "oaknut/feature_detection/feature_detection.hpp" then just call `detect_features` to get a bitset of features in a cross-platform manner. However, we also want the ability for advanced users to do custom things: Users can call more specific `detect_features_via_*` functions if they wish.

CPU feature detection is operating system specific, and some operating systems even have multiple methods. Here are a list of operating systems and implemented methods:

* Windows
  * [x] IsProcessorFeaturePresent
  * [x] Read EL1 registers from registry
* macOS
  * [x] sysctlbyname
* Linux
  * [x] ELF hwcaps
  * [x] CPUID interface
* FreeBSD
  * [x] ELF hwcaps
  * [x] CPUID interface
* NetBSD
  * [x] "machdep.cpu%d.cpu_id" sysctl
* OpenBSD
  * [x] CTL_MACHDEP interface

### CODE REVIEW

File: include/oaknut/feature_detection/feature_detection_freebsd.hpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+#endif
+#ifndef AT_HWCAP2
+#    define AT_HWCAP2 26
+#endif
+
+namespace oaknut {
+
+namespace detail {
+
+unsigned long getauxval(int aux)
```
Comment (merryhime) [inline include/oaknut/feature_detection/feature_detection_freebsd.hpp:22]: `inline`

**Comment by Wunkolo** [comment]
Can verify the current windows stuff works on my Thinkpad X13S:
```
CPU Features: FP ASIMD AES PMULL SHA1 SHA256 CRC32 LSE LRCPC DotProd
```

Here's the output of `Get-Item -Path HKLM:\HKEY_LOCAL_MACHINE\HARDWARE\DESCRIPTION\System\CentralProcessor\*` if it helps any on the windows side:

<details>
    <summary>Click to expand</summary>

```
    Hive: HKEY_LOCAL_MACHINE\HARDWARE\DESCRIPTION\System\CentralProcessor


Name                           Property
----                           --------
0                              Component Information : {0, 0, 0, 0...}
                               Identifier            : ARMv8 (64-bit) Family 8 Model D4B Revision   0
                               Configuration Data    :
                               VendorIdentifier      : Qualcomm Technologies Inc
                               ~MHz                  : 1440
                               CP 4020               : 1224979098931106066
                               CP 4021               : 16
                               CP 4028               : 271602696
                               CP 4029               : 0
                               CP 402C               : 0
                               CP 402D               : 0
                               CP 4030               : 4521192084017440
                               CP 4031               : 18874417
                               CP 4038               : 1048609
                               CP 4039               : 2097154
                               CP 403A               : 16
                               CP 4000               : 1091556528
                               CP 5801               : 2487533572
                               CP 4080               : 17593005005149
                               CP 4081               : 0
                               CP 4100               : 54043212695154688
                               CP 4510               : 4919057789357392127
                               ProcessorNameString   : Snapdragon (TM) 8cx Gen 3 @ 3.0 GHz
1                              Component Information : {0, 0, 0, 0...}
                               Identifier            : ARMv8 (64-bit) Family 8 Model D4B Revision   0
                               Configuration Data    :
                               VendorIdentifier      : Qualcomm Technologies Inc
                               ~MHz                  : 1440
                               CP 4020               : 1224979098931106066
                               CP 4021               : 16
                               CP 4028               : 271602696
                               CP 4029               : 0
                               CP 402C               : 0
                               CP 402D               : 0
                               CP 4030               : 4521192084017440
                               CP 4031               : 18874417
                               CP 4038               : 1048609
                               CP 4039               : 2097154
                               CP 403A               : 16
                               CP 4000               : 1091556528
                               CP 5801               : 2487533572
                               CP 4080               : 17593005005149
                               CP 4081               : 0
                               CP 4100               : 54043212695154688
                               CP 4510               : 4919057789357392127
                               ProcessorNameString   : Snapdragon (TM) 8cx Gen 3 @ 3.0 GHz
2                              Component Information : {0, 0, 0, 0...}
                               Identifier            : ARMv8 (64-bit) Family 8 Model D4B Revision   0
                               Configuration Data    :
                               VendorIdentifier      : Qualcomm Technologies Inc
                               ~MHz                  : 1440
                               CP 4020               : 1224979098931106066
                               CP 4021               : 16
                               CP 4028               : 271602696
                               CP 4029               : 0
                               CP 402C               : 0
                               CP 402D               : 0
                               CP 4030               : 4521192084017440
                               CP 4031               : 18874417
                               CP 4038               : 1048609
                               CP 4039               : 2097154
                               CP 403A               : 16
                               CP 4000               : 1091556528
                               CP 5801               : 2487533572
                               CP 4080               : 17593005005149
                               CP 4081               : 0
                               CP 4100               : 54043212695154688
                               CP 4510               : 4919057789357392127
                               ProcessorNameString   : Snapdragon (TM) 8cx Gen 3 @ 3.0 GHz
3                              Component Information : {0, 0, 0, 0...}
                               Identifier            : ARMv8 (64-bit) Family 8 Model D4B Revision   0
                               Configuration Data    :
                               VendorIdentifier      : Qualcomm Technologies Inc
                               ~MHz                  : 1440
                               CP 4020               : 1224979098931106066
                               CP 4021               : 16
                               CP 4028               : 271602696
                               CP 4029               : 0
                               CP 402C               : 0
                               CP 402D               : 0
                               CP 4030               : 4521192084017440
                               CP 4031               : 18874417
                               CP 4038               : 1048609
                               CP 4039               : 2097154
                               CP 403A               : 16
                               CP 4000               : 1091556528
                               CP 5801               : 2487533572
                               CP 4080               : 17593005005149
                               CP 4081               : 0
                               CP 4100               : 54043212695154688
                               CP 4510               : 4919057789357392127
                               ProcessorNameString   : Snapdragon (TM) 8cx Gen 3 @ 3.0 GHz
4                              Component Information : {0, 0, 0, 0...}
                               Identifier            : ARMv8 (64-bit) Family 8 Model D4C Revision   0
                               Configuration Data    :
                               VendorIdentifier      : Qualcomm Technologies Inc
                               ~MHz                  : 1747
                               CP 4020               : 1224979098931106066
                               CP 4021               : 16
                               CP 4028               : 271602696
                               CP 4029               : 0
                               CP 402C               : 0
                               CP 402D               : 0
                               CP 4030               : 4521192084017440
                               CP 4031               : 18874417
                               CP 4038               : 1048609
                               CP 4039               : 2097154
                               CP 403A               : 16
                               CP 4000               : 1091556544
                               CP 5801               : 2487533572
                               CP 4080               : 17593005005149
                               CP 4081               : 0
                               CP 4100               : 54043212695154688
                               CP 4510               : 4919057789357392127
                               ProcessorNameString   : Snapdragon (TM) 8cx Gen 3 @ 3.0 GHz
5                              Component Information : {0, 0, 0, 0...}
                               Identifier            : ARMv8 (64-bit) Family 8 Model D4C Revision   0
                               Configuration Data    :
                               VendorIdentifier      : Qualcomm Technologies Inc
                               ~MHz                  : 1747
                               CP 4020               : 1224979098931106066
                               CP 4021               : 16
                               CP 4028               : 271602696
                               CP 4029               : 0
                               CP 402C               : 0
                               CP 402D               : 0
                               CP 4030               : 4521192084017440
                               CP 4031               : 18874417
                               CP 4038               : 1048609
                               CP 4039               : 2097154
                               CP 403A               : 16
                               CP 4000               : 1091556544
                               CP 5801               : 2487533572
                               CP 4080               : 17593005005149
                               CP 4081               : 0
                               CP 4100               : 54043212695154688
                               CP 4510               : 4919057789357392127
                               ProcessorNameString   : Snapdragon (TM) 8cx Gen 3 @ 3.0 GHz
6                              Component Information : {0, 0, 0, 0...}
                               Identifier            : ARMv8 (64-bit) Family 8 Model D4C Revision   0
                               Configuration Data    :
                               VendorIdentifier      : Qualcomm Technologies Inc
                               ~MHz                  : 1747
                               CP 4020               : 1224979098931106066
                               CP 4021               : 16
                               CP 4028               : 271602696
                               CP 4029               : 0
                               CP 402C               : 0
                               CP 402D               : 0
                               CP 4030               : 4521192084017440
                               CP 4031               : 18874417
                               CP 4038               : 1048609
                               CP 4039               : 2097154
                               CP 403A               : 16
                               CP 4000               : 1091556544
                               CP 5801               : 2487533572
                               CP 4080               : 17593005005149
                               CP 4081               : 0
                               CP 4100               : 54043212695154688
                               CP 4510               : 4919057789357392127
                               ProcessorNameString   : Snapdragon (TM) 8cx Gen 3 @ 3.0 GHz
7                              Component Information : {0, 0, 0, 0...}
                               Identifier            : ARMv8 (64-bit) Family 8 Model D4C Revision   0
                               Configuration Data    :
                               VendorIdentifier      : Qualcomm Technologies Inc
                               ~MHz                  : 1747
                               CP 4020               : 1224979098931106066
                               CP 4021               : 16
                               CP 4028               : 271602696
                               CP 4029               : 0
                               CP 402C               : 0
                               CP 402D               : 0
                               CP 4030               : 4521192084017440
                               CP 4031               : 18874417
                               CP 4038               : 1048609
                               CP 4039               : 2097154
                               CP 403A               : 16
                               CP 4000               : 1091556544
                               CP 5801               : 2487533572
                               CP 4080               : 17593005005149
                               CP 4081               : 0
                               CP 4100               : 54043212695154688
                               CP 4510               : 4919057789357392127
                               ProcessorNameString   : Snapdragon (TM) 8cx Gen 3 @ 3.0 GHz
```
</details>

File: include/oaknut/feature_detection/id_registers.hpp [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+
+struct Pfr2Register {
+    std::uint64_t value;
+
+    constexpr unsigned MTEPERM() const { return detail::extract_field<0>(value); }
+    constexpr unsigned MTESTOREONLY() const { return detail::extract_field<4>(value); }
+    constexpr unsigned MTEFAR() const { return detail::extract_field<8>(value); }
+    // [12:31] reserved
+    constexpr unsigned FPMR() const { return detail::extract_field<32>(value); }
+    // [36:31] reserved
```
Comment (merryhime) [inline include/oaknut/feature_detection/id_registers.hpp:73]: // [36:63] reserved

**Comment by merryhime** [comment]
[ ] TODO: Consider whether correcting for https://github.com/freebsd/freebsd-src/commit/d81d009c6cfc6ff767f185cef129a2537cbbc5c4 is reasonable.

**Comment by merryhime** [comment]
@Wunkolo 

For future reference, afaik the registers in the Windows registry are labelled as per their encodings.

i.e.: op0:op1:crn:crm:op2

So MIDR is 0x4000.

**Comment by merryhime** [comment]
Complete.

**Pull request merged**